### ISSUE TITLE
Question: how should new cloud controller managers implement the HasClusterID, and related cluster ID, functionality?

### ISSUE THREAD

**Issue by elmiko** [kubernetes/kubernetes#128320]
_note from @elmiko, i am transferring this issue from its original location https://github.com/kubernetes/cloud-provider/issues/71_

originally posted by @guettli 

Looking at the basic_main.go

```go
	if !cloud.HasClusterID() {
		if config.ComponentConfig.KubeCloudShared.AllowUntaggedCloud {
			klog.Warning("detected a cluster without a ClusterID.  A ClusterID will be required in the future.  Please tag your cluster to avoid any future issues")
		} else {
			klog.Fatalf("no ClusterID found.  A ClusterID is required for the cloud provider to function properly.  This check can be bypassed by setting the allow-untagged-cloud option")
		}
	}
```
[basic_main.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/sample/basic_main.go#L90)

But where is ClusterID actually used?

Afaik allow-untagged-cloud is deprecated, but afaik ClusterID does not get used.

The current state is confusing.

Please explain how a ccm should implement a ClusterID.

/sig cloud-provider

**Comment by elmiko** [comment]
we discussed this issue at our last (23 Oct) sig meeting, and we felt this is a good topic of discussion and potential research topic for people looking to become more involved with sig cloud provider.

/triage accepted

**Comment by carlory** [comment]
I'm unsure whether it is related to the project under the sig-multicluster. https://github.com/kubernetes-sigs/cluster-inventory-api

It allows a cluster provider or admin to define a cluster's inventory like properties, cluster version and etc.

**Comment by carlory** [comment]
@dixudx (clusternet maintainer) and @RainbowMango (karmada maintainer) may have more insights on this. clusternet and karmada focus on the multi-cluster management, which is a hot topic in the cloud-native community.

**Comment by RainbowMango** [comment]
@carlory I guess it's nothing to do with [the cluster-inventory-api](https://github.com/kubernetes-sigs/cluster-inventory-api), nor [the about api](https://github.com/kubernetes-sigs/about-api), as this was introduced 7 years ago, earlier than them.

It seems no concrete use case for it yet. Why not deprecate it?

**Comment by elmiko** [comment]
> I'm unsure whether it is related to the project under the sig-multicluster. https://github.com/kubernetes-sigs/cluster-inventory-api

@carlory as @RainbowMango mentioned, i do not believe this is related to the cluster-inventory-api as this function has been part of the cloud-provider interface for many years. we (the sig) were unsure about its usage as well, and thought this seemed like a good issue for investigation.

> It seems no concrete use case for it yet. Why not deprecate it?

i think this is the root of the original question.

**Comment by carlory** [comment]
cc @guettli @elmiko @dims The [aws cloud provider](https://github.com/kubernetes/cloud-provider-aws) is still using the `clusterID`. And the `clusterID` property is marked as required. why cluster id is required for the AWS cloud provider? please refer to [Make the cluster-id tag required for aws](https://github.com/kubernetes/kubernetes/issues/48954): the `clusterID` is required for any cloud provider unless `--allow-untagged-cloud=true` is passed. 

There is another issue. [Change mechanism of zone selection for dynamic provisioning](https://github.com/kubernetes/kubernetes/issues/48504) It is related to cluster identification. 

In the past, some in-tree cloud provider volume plugins used `clusterID`. As CSI migration is completed, the `clusterID` is not used in in-tree volume plugins.

- https://github.com/kubernetes/kubernetes/pull/52322

/sig storage

cc @msau42 @jsafrane @gnufied 

In conclusion, in my opinion (I may be wrong), the `clusterID` is required for all cloud providers. For backward compatibility, it also allows the `--allow-untagged-cloud=true` to disable the `clusterID` requirement for the existing cluster without the `clusterID` tag created by another cloud provider. The final goal is to make all cloud providers support the `clusterID` tag. However, the goal has not been achieved yet.

The requirement of `clusterID` may change over time. Before making a decision, it is necessary to discuss it with primary cloud providers (AWS, GCP, Azure, Alibaba, etc.).

**Comment by dims** [comment]
cc @kmala @cartermckinnon

**Comment by cartermckinnon** [comment]
The AWS cloud provider uses `clusterID` to disambiguate the resources that it manages: https://github.com/kubernetes/cloud-provider-aws/blob/e101f87f97698ac7df96f3d32b469b3c3c5511b6/pkg/providers/v1/tags.go#L58-L61

That allows you to run multiple clusters in the same AWS account/VPC/subnet.

For example, when creating a load balancer the `clusterID` is used to generate the target group name: https://github.com/kubernetes/cloud-provider-aws/blob/e101f87f97698ac7df96f3d32b469b3c3c5511b6/pkg/providers/v1/aws_loadbalancer.go#L499

Other resources are tagged with the `clusterID`, such as route tables: https://github.com/kubernetes/cloud-provider-aws/blob/e101f87f97698ac7df96f3d32b469b3c3c5511b6/pkg/providers/v1/aws_routes.go#L52

**Comment by carlory** [comment]
The following projects are related to this issue:
1. kube-controller-manager
2. cloud-controller-manager
3. cloud-provider-xxx (e.g. cloud-provider-gcp, cloud-provider-aws, cloud-provider-azure, etc.)

cloud-controller-manager was a effort to move cloud provider specific code out of kube-controller-manager, but it is replaced by `cloud-provider-xxx` since 1.20. 

`clusterID` is the hard requirement for aws cloud provider, so they raise a PR to extend the `cloud.Interface` interface to include `HasClusterID` method. A side effect of this change is enforcing all cloud providers to implement this method, even if some providers do not need it. 

```
func (az *Cloud) HasClusterID() bool {
	return true
}
func (cs *CSCloud) HasClusterID() bool {
	return true
}
// HasClusterID returns true if the cluster has a clusterID
func (os *OpenStack) HasClusterID() bool {
	return true
}
```

I'm not sure if this change is agreed by all cloud providers, I believe it should have a discussion in the sig-cloud-provider before making this change.

If this change or enforcement is not agreed by all cloud providers, I have a proposal to remove this method from `cloud.Interface` interface, we do implement the same goal by doing it in a different way like below:

```go
// cloudprovider.go
func init() {
	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
        // read clusterID from config or env
        // it depends on the specific cloud provider
        if clusterID == "" {
            return nil, fmt.Errorf("clusterID is required")
        }
        return &Cloud{}
    })    
}

// main.go in cloud-provider-xxx and cloud-controller-manager
func cloudInitializer(config *config.CompletedConfig) cloudprovider.Interface {
	cloudConfig := config.ComponentConfig.KubeCloudShared.CloudProvider

	// initialize cloud provider with the cloud provider name and config file provided
	cloud, err := cloudprovider.InitCloudProvider(cloudConfig.Name, cloudConfig.CloudConfigFile)
	if err != nil {
		klog.Fatalf("Cloud provider could not be initialized: %v", err)
	}
	if cloud == nil {
		klog.Fatalf("Cloud provider is nil")
	}

	// if !cloud.HasClusterID() {
	// 	if config.ComponentConfig.KubeCloudShared.AllowUntaggedCloud {
	// 		klog.Warning("detected a cluster without a ClusterID.  A ClusterID will be required in the future.  Please tag your cluster to avoid any future issues")
	// 	} else {
	// 		klog.Fatalf("no ClusterID found.  A ClusterID is required for the cloud provider to function properly.  This check can be bypassed by setting the allow-untagged-cloud option")
	// 	}
	// }
	return cloud
}

```

What do you think?

**Comment by cartermckinnon** [comment]
Generally agree, but removing it would require changes by every cloud.Interface maintainer (whether they need it or not). We can’t get a complete picture of usage, because folks may keep their impl private. It’s very simple to implement as a no-op, so I don’t see much benefit to the breakage

**Comment by elmiko** [comment]
thanks for all the replies here, it definitely helps.

it seems like we definitely need to keep this function around, but perhaps we should add more documentation to the code in the cloud-provider framework. i think just letting implementers know that this field can be used for cloud-specific identifiers but that kubernetes does not use the value outside of that (if i'm understanding correctly) would be helpful.

**Comment by elmiko** [comment]
i am updating the title of this issue to reflect our new understanding.

basically, we should update the code documentation to give better guidance to cloud controller implementers. in specific i think we should add more information to 2 places in the cloud-provider framework:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/cloud.go#L67-L68
this is the location of the `HasClusterID` function, i think we should explain a little more about cluster id and its relationship to provider-specific behavior. perhaps with an example of why to use and why not to use.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/sample/basic_main.go#L90-L96
this is the location for the example main, which contains a conditional on the `HasClusterID`, and can produce an error if it is not found. we should add more documentation to this to describe why clusterid might be needed or not, and that if a provider does not have a clusterid then either they should avoid this logic for make more use of the `allow-untagged-cloud` option.

i think adding more guidance here with help in the future, and hopefully answer @guettli 's questions as well.

**Comment by elmiko** [comment]
/title cloud-provider HasClusterID related functionality should have better documentation in the code

/help-wanted

**Comment by elmiko** [comment]
we had a discussion about this issue at today's sig cloud provider meeting, we are going to work towards creating better documentation around ClusterID and its usage. thanks for all the input here folks!

**Comment by Ogu1208** [comment]
Hi! I’m interested in contributing to this issue. Can I work on it?

**Comment by elmiko** [comment]
i am coordinating with @Ogu1208 in slack about this issue.
